plugins/base/src/main/kotlin/signatures/KotlinSignatureProvider.kt

```diff
@@ -0,0 +1,112 @@
+import org.jetbrains.dokka.links.sureClassNames
+import org.jetbrains.dokka.model.*
+import org.jetbrains.dokka.model.Annotation
+import org.jetbrains.dokka.model.Enum
+import org.jetbrains.dokka.model.Function
+import org.jetbrains.dokka.pages.ContentKind
+import org.jetbrains.dokka.pages.ContentNode
+import org.jetbrains.dokka.pages.PlatformData
+import org.jetbrains.dokka.utilities.DokkaLogger
+
+class KotlinSignatureProvider(ctcc: CommentsToContentConverter, logger: DokkaLogger) : SignatureProvider {
+    private val contentBuilder = PageContentBuilder(ctcc, this, logger)
+
+    override fun signature(documentable: Documentable): ContentNode = when (documentable) {
+        is Function -> signature(documentable)
+        is Classlike -> signature(documentable)
+        is TypeParameter -> signature(documentable)
+        else -> throw NotImplementedError(
+            "Cannot generate signature for ${documentable::class.qualifiedName} ${documentable.name}"
+        )
+    }
+
+    private fun signature(c: Classlike) = contentBuilder.contentFor(c, ContentKind.Symbol) {
+        platformText(c.visibility) { it.externalDisplayName + " " }
+        if (c is Class) {
+            text(c.modifier.toString() + " ")
+        }
+        when (c) {
+            is Class -> text("class ")
+            is Interface -> text("interface ")
+            is Enum -> text("enum ")
+            is Object -> text("object ")
+            is Annotation -> text("annotation class ")
+        }
+        text(c.name!!)
+        if (c is WithSupertypes) {
+            c.supertypes.map { (p, dris) ->
+                list(dris, prefix = " : ", platformData = setOf(p)) {
+                    link(it.sureClassNames, it, platformData = setOf(p))
+                }
+            }
+        }
+    }
+
+    private fun signature(f: Function) = contentBuilder.contentFor(f, ContentKind.Symbol) {
+        platformText(f.visibility) { it.externalDisplayName + " " }
+        text(f.modifier.toString().toLowerCase() + " ")
+        text("fun ")
+        f.receiver?.also {
+            type(it.type)
+            text(".")
+        }
+        link(f.name, f.dri)
+        list(f.generics, prefix = "<", suffix = ">") {
+            +buildSignature(it)
+        }
+        text("(")
+        list(f.parameters) {
+            link(it.name!!, it.dri)
+            text(": ")
+            type(it.type)
+        }
+        text(")")
+        val returnType = f.type
+        if (!f.isConstructor && returnType.constructorFqName != Unit::class.qualifiedName) {
+            text(": ")
+            type(returnType)
+        }
+    }
+
+    private fun signature(t: TypeParameter) = contentBuilder.contentFor(t, ContentKind.Symbol) {
+        link(t.name, t.dri)
+        list(t.bounds, prefix = " : ") {
+            signatureForProjection(it)
+        }
+    }
+
+    private fun PageContentBuilder.DocumentableContentBuilder.signatureForProjection(p: Projection): Unit = when (p) {
```

Comment:
I think you can just wrap this in `contentBuilder.contentFor(){}` and then use it with plus: `+signatureForProjection(it)`, like `buildSignature`